### PR TITLE
Create circuit schematics

### PR DESCRIPTION
**Pull request by maxcapodi78**
Three new methods in Circuit class to create automatic schematics for LNA, TDR and AMI analysis starting from a touchstone file

### CODE REVIEW

File: pyaedt/application/Analysis3DLayout.py [hunk 1]
```diff
@@ -236,40 +236,55 @@ def export_mesh_stats(self, setup_name, variation_string="", mesh_path=None):
         return mesh_path
 
     @pyaedt_function_handler()
-    def get_all_return_loss_list(self, excitation_names=[], excitation_name_prefix=""):
-        """Retrieve a list of all return losses for a list of excitations.
+    def get_all_return_loss_list(
+        self, excitation_names=None, excitation_name_prefix="", math_formula="", net_list=None
+    ):
+        """Retrieve a list of all return losses for a list of exctitations.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:242]:
```suggestion
        """Get a list of all return losses for a list of excitations.
```
Use the same verb as the method name for consistency/clarity.

File: pyaedt/application/Analysis3DLayout.py [hunk 2]
```diff
@@ -236,40 +236,55 @@ def export_mesh_stats(self, setup_name, variation_string="", mesh_path=None):
...
 
         Parameters
         ----------
         excitation_names : list, optional
-            List of excitations. The default is ``[]``, in which case
+            List of excitations. The default is ``None``, in which case
             the return losses for all excitations are to be provided.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:248]:
```suggestion
            the return losses for all excitations are provided.
```

File: pyaedt/application/Analysis3DLayout.py [hunk 3]
```diff
@@ -236,40 +236,55 @@ def export_mesh_stats(self, setup_name, variation_string="", mesh_path=None):
...
-            For example, ``["1", "2"]``.
+            For example ``["1", "2"]``.
         excitation_name_prefix : string, optional
-             Prefix to add to the excitation names. The default is ``""``.
+             Prefix to add to the excitation names. The default is ``""``,
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:255]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which case all parameters are returned.
```

File: pyaedt/application/Analysis3DLayout.py [hunk 4]
```diff
@@ -236,40 +236,55 @@ def export_mesh_stats(self, setup_name, variation_string="", mesh_path=None):
...
 
         Returns
         -------
-        list
+        list of str
             List of strings representing the return losses of the excitations.
-            For example, ``["S(1, 1)", "S(2, 2)"]``.
+            For example ``["S(1, 1)", S(2, 2)]``.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:261]:
```suggestion
            For example, ``["S(1, 1)", S(2, 2)]``.
```
The comma should be there to separate the introductory clause from the code entity.

File: pyaedt/application/Analysis3DLayout.py [hunk 5]
```diff
@@ -236,40 +236,55 @@ def export_mesh_stats(self, setup_name, variation_string="", mesh_path=None):
...
+                else:
+                    spar.append("S({},{})".format(i, i))
         return spar
 
     @pyaedt_function_handler()
-    def get_all_insertion_loss_list(self, trlist=[], reclist=[], tx_prefix="", rx_prefix=""):
+    def get_all_insertion_loss_list(
+        self, trlist=None, reclist=None, tx_prefix="", rx_prefix="", math_formula="", net_list=[]
+    ):
         """Retrieve a list of all insertion losses from two lists of excitations (driver and receiver).
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:288]:
```suggestion
        """Get a list of all insertion losses from two lists of excitations (driver and receiver).
```

File: pyaedt/application/Analysis3DLayout.py [hunk 6]
```diff
@@ -283,102 +298,160 @@ def get_all_insertion_loss_list(self, trlist=[], reclist=[], tx_prefix="", rx_pr
             Prefix to add to driver names. For example, ``"DIE"``. The default is ``""``.
         rx_prefix : str, optional
             Prefix to add to receiver names. For example, ``"BGA"``. The default is ``""``.
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:304]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which
            case all parameters are returned.
```

File: pyaedt/application/Analysis3DLayout.py [hunk 7]
```diff
@@ -283,102 +298,160 @@ def get_all_insertion_loss_list(self, trlist=[], reclist=[], tx_prefix="", rx_pr
...
+                if math_formula:
+                    spar.append("{}(S({},{}))".format(math_formula, i, j))
+                else:
+                    spar.append("S({},{})".format(i, j))
         return spar
 
     @pyaedt_function_handler()
-    def get_next_xtalk_list(self, trlist=[], tx_prefix=""):
+    def get_next_xtalk_list(self, trlist=[], tx_prefix="", math_formula="", net_list=None):
         """Retrieve a list of all the near end XTalks from a list of excitations (driver and receiver).
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:352]:
```suggestion
        """Get a list of all the near end XTalks from a list of excitations (driver and receiver).
```

File: pyaedt/application/Analysis3DLayout.py [hunk 8]
```diff
@@ -283,102 +298,160 @@ def get_all_insertion_loss_list(self, trlist=[], reclist=[], tx_prefix="", rx_pr
...
         trlist : list, optional
-            List of drivers. The default is ``[]``. For example, ``["1", "2", "3"]``.
+            List of drivers. The default is ``[]``. For example,
+            ``["1", "2", "3"]``.
         tx_prefix : str, optional
             Prefix to add to driver names. For example, ``"DIE"``.  The default is ``""``.
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:364]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which case
            all parameters are returned.
```

File: pyaedt/application/Analysis3DLayout.py [hunk 9]
```diff
@@ -283,102 +298,160 @@ def get_all_insertion_loss_list(self, trlist=[], reclist=[], tx_prefix="", rx_pr
...
+            Prefix for receiver names. For examples, ``"BGA"`` The default is ``""``.
         skip_same_index_couples : bool, optional
             Whether to skip driver and receiver couples with the same index position.
             The default is ``True``, in which case the drivers and receivers
             with the same index position are considered insertion losses and
             excluded from the list.
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:424]:
```suggestion
            List of nets to filter the output. The defautl is ``None``, in which case all
            parameters are returned.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 10]
```diff
@@ -381,7 +381,9 @@ def get_all_sparameter_list(self, excitation_names=[]):
         return spar
 
     @pyaedt_function_handler()
-    def get_all_return_loss_list(self, excitation_names=None, excitation_name_prefix=""):
+    def get_all_return_loss_list(
+        self, excitation_names=None, excitation_name_prefix="", math_formula="", net_list=None
+    ):
         """Retrieve a list of all return losses for a list of exctitations.
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:387]:
```suggestion
        """Geta list of all return losses for a list of excitations.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 11]
```diff
@@ -392,6 +394,10 @@ def get_all_return_loss_list(self, excitation_names=None, excitation_name_prefix
             For example ``["1", "2"]``.
         excitation_name_prefix : string, optional
              Prefix to add to the excitation names. The default is ``""``,
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:400]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which case
            all parameters are returned.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 12]
```diff
@@ -413,11 +419,17 @@ def get_all_return_loss_list(self, excitation_names=None, excitation_name_prefix
...
+                else:
+                    spar.append("S({},{})".format(i, i))
         return spar
 
     @pyaedt_function_handler()
-    def get_all_insertion_loss_list(self, trlist=None, reclist=None, tx_prefix="", rx_prefix=""):
+    def get_all_insertion_loss_list(
+        self, trlist=None, reclist=None, tx_prefix="", rx_prefix="", math_formula="", net_list=[]
+    ):
         """Retrieve a list of all insertion losses from two lists of excitations (driver and receiver).
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:433]:
```suggestion
        """Get a list of all insertion losses from two lists of excitations (driver and receiver).
```

File: pyaedt/application/AnalysisNexxim.py [hunk 13]
```diff
@@ -431,6 +443,10 @@ def get_all_insertion_loss_list(self, trlist=None, reclist=None, tx_prefix="", r
             Prefix to add to driver names. For example, ``"DIE"``. The default is ``""``.
         rx_prefix : str, optional
             Prefix to add to receiver names. For example, ``"BGA"``. The default is ``""``.
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:449]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which case
            all parameters are returned.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 14]
```diff
@@ -443,25 +459,41 @@ def get_all_insertion_loss_list(self, trlist=None, reclist=None, tx_prefix="", r
...
+                if math_formula:
+                    spar.append("{}(S({},{}))".format(math_formula, i, j))
+                else:
+                    spar.append("S({},{})".format(i, j))
         return spar
 
     @pyaedt_function_handler()
-    def get_next_xtalk_list(self, trlist=[], tx_prefix=""):
+    def get_next_xtalk_list(self, trlist=[], tx_prefix="", math_formula="", net_list=None):
         """Retrieve a list of all the near end XTalks from a list of excitations (driver and receiver).
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:497]:
```suggestion
        """Get a list of all the near end XTalks from a list of excitations (driver and receiver).
```

File: pyaedt/application/AnalysisNexxim.py [hunk 15]
```diff
@@ -471,6 +503,10 @@ def get_next_xtalk_list(self, trlist=[], tx_prefix=""):
             ``["1", "2", "3"]``.
         tx_prefix : str, optional
             Prefix to add to driver names. For example, ``"DIE"``.  The default is ``""``.
+        math_formula : str, optional
+            One of the available AEDT mathematical formulas to apply. For example, ``abs, dB``.
+        net_list : list, optional
+            List of nets to filter the output. If None, all parameters will be returned.
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:509]:
```suggestion
            List of nets to filter the output. The default is ``None``, in which case
            all parameters are returned.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 16]
```diff
@@ -487,14 +523,27 @@ def get_next_xtalk_list(self, trlist=[], tx_prefix=""):
...
+        self,
+        trlist=None,
+        reclist=None,
+        tx_prefix="",
+        rx_prefix="",
+        skip_same_index_couples=True,
+        math_formula="",
+        net_list=None,
+    ):
         """Retrieve a list of all the far end XTalks from two lists of exctitations (driver and receiver).
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:547]:
```suggestion
        """Get a list of all the far end XTalks from two lists of excitations (driver and receiver).
```

File: pyaedt/circuit.py [hunk 17]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+        probe_p_pins,
+        probe_ref_pins=None,
+        termination_pins=None,
+        differential=True,
+        rise_time=30,
+        use_convolution=True,
+        analyze=True,
+        design_name="LNA",
+    ):
+        """Create a new schematic from a touchstone file and automatically setup a TDR transient analysis.
```
Comment (PipKat) [inline pyaedt/circuit.py:1705]:
```suggestion
        """Create a schematic from a Touchstone file and automatically setup a TDR transient analysis.
```
Create means "new" so the word "new" isn't needed.

File: pyaedt/circuit.py [hunk 18]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+
+        Parameters
+        ----------
+        touchstone : str
+            Full path to the sNp file.
+        probe_p_pins : list
+            Pins to attach to the probe components.
+        probe_ref_pins : list, optional
+            Reference pins to attach to probe components. Valid only in differentila con
```
Comment (PipKat) [inline pyaedt/circuit.py:1714]:
```suggestion
            Reference pins to attach to probe components. The default is ``None``.
            This parameter is valid only in differential con
```
The last sentence is incomplete.

File: pyaedt/circuit.py [hunk 19]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+        termination_pins : list
+            Pins to be terminated.
```
Comment (PipKat) [inline pyaedt/circuit.py:1716]:
```suggestion
            Pins to terminated. The default is ``None``.
```

File: pyaedt/circuit.py [hunk 20]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+
+        Parameters
+        ----------
+        touchstone : str
+            Full path to the sNp file.
+        probe_p_pins : list
+            Pins to attach to the probe components.
+        probe_ref_pins : list, optional
+            Reference pins to attach to probe components. Valid only in differentila con
+        termination_pins : list
```
Comment (PipKat) [inline pyaedt/circuit.py:1715]:
```suggestion
        termination_pins : list, optional
```

File: pyaedt/circuit.py [hunk 21]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+            Pins to be terminated.
+        differential : bool, optional
+            Whether if the buffers are single ended or differential. Default is ``True`` to use differential buffer.
```
Comment (PipKat) [inline pyaedt/circuit.py:1718]:
```suggestion
            Whether the buffers are differential. The default is ``True``. If ``False``, the
            pins are single ended.
```

File: pyaedt/circuit.py [hunk 22]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+        rise_time : float, int, optional
+            Rise time of the input pulse in picoseconds. Default is ``30``.
+        use_convolution : bool, optional
+            Whether if use or not convolution for the touchstone file. If ``False`` default State-Space will be used.
+        analyze : bool
+             Whether if automatically assign differential pairs or not.  Default value is ``False``.
+        design_name : str, optional
+            New schematic name. Default is ``"LNA"``.
```
Comment (PipKat) [inline pyaedt/circuit.py:1726]:
```suggestion
        rise_time : float, int, optional
            Rise time of the input pulse in picoseconds. The default is ``30``.
        use_convolution : bool, optional
            Whether to use convolution for the Touchstone file. The default is ``True``.
            If ``False``, state-space is used.
        analyze : bool
             Whether to automatically assign differential pairs. The default is ``False``.
        design_name : str, optional
            New schematic name. The default is ``"LNA"``.
```

File: pyaedt/circuit.py [hunk 23]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+
+        Returns
+        -------
+        (bool, list, list)
+            First argument is ``True`` if succeeded.
+            Second and third argument are respectly differential and common mode pairs name.
```
Comment (PipKat) [inline pyaedt/circuit.py:1863]:
```suggestion
        """Create a schematic from a Touchstone file and automatically set up an LNA analysis.
        
        This method optionally assigns differential pairs automatically based on name pattern.

        Parameters
        ----------
        touchstone : str
            Full path to the sNp file.
        start_frequency : int, float, optional
            Start frequency in GHz of the LNA setup. The default is ``0``.
        stop_frequency  : int, float, optional
            Stop frequency in GHz of the LNA setup. The default is ``30``.
        auto_assign_diff_pairs : bool
            Whether to automatically assign differential pairs. The default is ``False``.
        separation : str, optional
            Character to use to separate port names. The default is ``"."``.
        pattern : list, optional
            Port name pattern. The default is ``["component", "pin", "net"]``.
        analyze : bool
             Whether to automatically assign differential pairs. The default is ``False``.
        design_name : str, optional
            New schematic name. The default is ``"LNA"``.

        Returns
        -------
        (bool, list, list)
            First argument is ``True`` if succeeded.
            Second and third argument are respectly differential and common mode pairs name.
```

File: pyaedt/circuit.py [hunk 24]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+        rx_refs,
+        use_ibis_buffer=True,
+        differential=True,
+        bit_pattern=None,
+        unit_interval=None,
+        use_convolution=True,
+        analyze=False,
+        design_name="AMI",
+    ):
+        """Create a new schematic from a touchstone file and automatically setup an IBIS AMI analysis.
```
Comment (PipKat) [inline pyaedt/circuit.py:1949]:
```suggestion
        """Create a schematic from a Touchstone file and automatically set up an IBIS-AMI analysis.
```
Hyphenated "IBIS-AMI" based on a Google search for the term.

File: pyaedt/circuit.py [hunk 25]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+        bit_pattern : str, optional
+            Ibis bit pattern.
+        unit_interval : str, optional
+            Unit interval of the bit pattern with units.
+        use_convolution : bool, optional
+            Whether if use or not convolution for the touchstone file. If ``False`` default State-Space will be used.
+        analyze : bool
+             Whether if automatically assign differential pairs or not.  Default value is ``False``.
+        design_name : str, optional
+            New schematic name. Default is ``"LNA"``.
```
Comment (PipKat) [inline pyaedt/circuit.py:1984]:
```suggestion
        touchstone : str
            Full path to the sNp file.
        ibis_ami : str
            Full path to the IBIS file.
        component_name : str
            Component name in the IBIS file to assign to components.
        tx_buffer_name : str
            Transmission buffer name.
        rx_buffer_name : str
            Receiver buffer name
        tx_pins : list
            Pins to assign the transmitter IBIS.
        tx_refs : list
            Reference pins to assign the transmitter IBIS. This parameter is only used in
            a differential configuration.
        rx_pins : list
            Pins to assign the receiver IBIS.
        rx_refs : list
            Reference pins to assign the receiver IBIS. This parameter is only used
            in a differential configuration.
        use_ibis_buffer : bool, optional
            Whether to use the IBIS buffer. The default is ``True``. If ``False``, pins are used.
        differential : bool, optional
            Whether the buffers are differential. The default is ``True``. If ``False``,
            the buffers are single-ended.
        bit_pattern : str, optional
            IBIS bit pattern.
        unit_interval : str, optional
            Unit interval of the bit pattern.
        use_convolution : bool, optional
            Whether to use convolution for the Touchstone file. The default is
            ``True``. If ``False``, state-space is used.
        analyze : bool
             Whether to automatically assign differential pairs. The default is ``False``.
        design_name : str, optional
            New schematic name. The default is ``"LNA"``.
```

File: pyaedt/circuit.py [hunk 26]
```diff
@@ -1687,3 +1688,404 @@ def import_edb_in_circuit(self, edb_path):
...
+
+
+        Returns
+        -------
+        (bool, list, list)
+            First argument is ``True`` if succeeded.
+            Second and third argument are respectly tx and rx mode probe names.
```
Comment (PipKat) [inline pyaedt/circuit.py:1991]:
```suggestion
            First argument is ``True`` if successful.
            Second and third arguments are respectively the names of the tx and rx mode probes.
```

File: pyaedt/modeler/circuits/object3dcircuit.py [hunk 27]
```diff
@@ -150,6 +152,8 @@ def connect_to_component(self, component_pin, page_name=None, use_wire=False, wi
             Wire name used only when user_wire is ``True``. Default value is ``""``.
```
Comment (PipKat) [inline pyaedt/modeler/circuits/object3dcircuit.py:152]:
```suggestion
            Wire name used only when ``user_wire=True``. The default is ``""``.
```

File: pyaedt/modeler/circuits/object3dcircuit.py [hunk 28]
```diff
@@ -150,6 +152,8 @@ def connect_to_component(self, component_pin, page_name=None, use_wire=False, wi
...
         clearance_units : int, optional
             Number of snap units (100mil each) around the object to overcome pins and wires.
+        page_port_angle : int, optional
+            Page port angle on the source pin. If None, the angle is automatically computed.
```
Comment (PipKat) [inline pyaedt/modeler/circuits/object3dcircuit.py:156]:
```suggestion
            Page port angle on the source pin. The default is ``None``, in which case
            the angle is automatically computed.
```

**Review by Samuelopez-ansys — DISMISSED** [review]

**Review by Samuelopez-ansys — DISMISSED** [review]

File: pyaedt/application/Analysis3DLayout.py [hunk 29]
```diff
@@ -148,7 +148,7 @@ def excitations(self):
         Returns
         -------
         list
-            List of excitation names. Excitations with multiple modes will return one
+            Excitation list. Excitations with multiple modes will return one
```
Comment (PipKat) [inline pyaedt/application/Analysis3DLayout.py:151]:
```suggestion
            Excitation list. Excitations with multiple modes return one
```
Use present tense.

**Review by Samuelopez-ansys — APPROVED** [review]

**Review by Samuelopez-ansys — APPROVED** [review]

**Pull request merged**